pkg/kubefed2/federate/federate.go

```diff
@@ -141,29 +154,40 @@ func (j *federateResource) Run(cmdOut io.Writer, config util.FedConfig) error {
 		Name:      j.resourceName,
 	}
 
-	resources, err := GetFedResources(hostConfig, qualifiedTypeName, qualifiedResourceName, j.outputYAML)
+	artifacts, err := GetFederateArtifacts(hostConfig, qualifiedTypeName, qualifiedResourceName, j.enableType, j.outputYAML)
```

Comment:
I'm not sure I understand why it is necessary to collect 'artifacts' required by enable vs just ensuring that a type is enabled before attempting federation.  The strategy as proposed appears to avoid treating the enable functionality as a black box, which will have negative implications when inevitable changes need to be made to that functionality.  Is there a good reason not to do the following?

```golang
if enableType {
  // This should be a function on the enable side of things.  The lack of a returned error indicates that enable was successful.
  ensureEnabled(...)
}

// Whatever operations are necessary to federate the type, including looking up the type config.
...
```

Reply:
1. For the flow you have suggested - I have kept any modification to the API at the end which minimises partial updates to the API in case of other failures, which includes enabling the type. I think, its better to keep it this way. Let me know if you think otherwise.

2. The artifacts which are collected are not the ones computed solely for enable. These are anyways computed for federate functionality. The exception is `apiResource`, which is computed but need not be passed further if not for `enableType`. `apiResource` infact can be fetched again from `typeConfig`, rather then being passed as another parameter.

3. About treating the enable as a black box - I did choose not to repeat the operations which are common to both (`lookup api type and generate the typeconfig`), which would be wasteful especially in the next PR (enable ns with content). I think its better this way and the coupling between the two functionalities is really really small. If you think strongly otherwise, I would not mind changing this such that only the two main enable functions (`GetResources` and `CreateResources`) are utilised in federate.

Reply:
1. Should `federate` commit to duplicating the error handling of `enable`?
2. Does the expense of an api call or two justify exposing `federate` to the internal details of `enable`?
3. As per my preliminary comments on #741, `enable` should be required at most once per type vs the proposed approach of once per resource.

I feel strongly that `enable` and `federate` are orthogonal concerns, and that `federate` should not be dependent on the internal details of `enable` until such time as performance is determined to be a concern.

Reply:
> * Are you proposing to minimize partial updates in `federate` but not in `enable`?

Did not get you? Enable already follows this flow, resources are created in the end in enable, isn't it?
> * Does the expense of some api calls justify exposing `federate` to the internal details of `enable`?

If you talking about my point 2 above, I was talking about artifacts, they are not really for enable, but needed for federate anyways. If you talking about my point 3. I already mentioned that I am quite ok to update such that no changes are made to enable and federate does not use anything internal ( eg `typeResources` ) from enable.
> * As per my preliminary comments on #741, `enable` should be required at most once per type vs the proposed approach of once per resource.

Thats understood. I was waiting for this PR to be complete before updating #741.

I have updated this PR accordingly, PTAL!